zerver/openapi/javascript_examples.py

```diff
@@ -0,0 +1,153 @@
+
+from typing import Dict, Any, Callable, Set, List
+
+import json
+import os
+import sys
+import subprocess
+
+from functools import wraps
+
+from zerver.openapi.openapi import validate_against_openapi_schema
+
+from zulip import Client
+
+TEST_FUNCTIONS = dict()  # type: Dict[str, Callable[..., None]]
+REGISTERED_TEST_FUNCTIONS = set()  # type: Set[str]
+CALLED_TEST_FUNCTIONS = set()  # type: Set[str]
+
+def openapi_test_function(endpoint: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """This decorator is used to register an openapi test function with
+    its endpoint. Example usage:
+
+    @openapi_test_function("/messages/render:post")
+    def ...
+    """
+    def wrapper(test_func: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(test_func)
+        def _record_calls_wrapper(*args: Any, **kwargs: Any) -> Any:
+            CALLED_TEST_FUNCTIONS.add(test_func.__name__)
+            return test_func(*args, **kwargs)
+
+        REGISTERED_TEST_FUNCTIONS.add(test_func.__name__)
+        TEST_FUNCTIONS[endpoint] = _record_calls_wrapper
+
+        return _record_calls_wrapper
+    return wrapper
+
+def run_js_code(js_code: str) -> List[Dict[str, Any]]:
+    """Here, we run the JavaScript examples and return
+    a list of all the responses from the code example
+    as json."""
+
+    responses_str = subprocess.check_output(
```

Comment:
I wrote this in a bigger comment, but it seems needlessly complex that we're code-generating JS here and shelling out to `node` instead of just having a human-readable JS file that we run tests against directly from `node`.  And then that reduces the problem to having the ancillary pieces here (document generation, etc.) just sweep the JS file for code snippets, which is really no more difficult.

Reply:
I think there's some benefit to reading from the code-generated result, because it's more of an end-to-end test of what will actually appear in the documentation.  At least, that's why we use the approach for the curl documentation, though that's a bit of a special case due to exactly how it gets generated.

The other problem is data setup -- we don't have as convenient a set of options for setting up the test database in the state we want in JS.  But maybe that's something we can work around in an acceptable way.

Reply:
> I think there's some benefit to reading from the code-generated result, because it's more of an end-to-end test of what will actually appear in the documentation. At least, that's why we use the approach for the curl documentation, though that's a bit of a special case due to exactly how it gets generated.

I don't think we intend to do code-generation here.  For **curl** we do actually code-generate examples, but for **python** and **javascript** we just hand-write it, since it gives us a chance to add comments, deal with quirks, etc., without having to write complicated code.  Even the curl code generation--which, in principle, is pretty straightforward--is pretty hard code to read.

As far as reading the code for docs, that's trivial, and the three of us have talked this through.


 
> The other problem is data setup -- we don't have as convenient a set of options for setting up the test database in the state we want in JS. But maybe that's something we can work around in an acceptable way.

That's a good point, but I think we can work around this.  We essentially want to do what we do with the Casper tests--for 90% of endpoints, the fixture data is already what you need, and then for certain things you just want to sequence the operations (do an add before the remove, etc.).